src/daemon/daemon.cpp

```diff
@@ -1544,119 +1722,72 @@ try // clang-format on
     mpl::ClientLogger<InfoReply, InfoRequest> logger{mpl::level_from(request->verbosity_level()), *config->logger,
                                                      server};
     InfoReply response;
+    InstanceSnapshotsMap instance_snapshots_map;
     bool have_mounts = false;
     bool deleted = false;
-    auto fetch_info = [&](VirtualMachine& vm) {
-        const auto& name = vm.vm_name;
-        auto info = response.add_info();
-        auto present_state = vm.current_state();
-        info->set_name(name);
-        if (deleted)
-        {
-            info->mutable_instance_status()->set_status(mp::InstanceStatus::DELETED);
-        }
-        else
-        {
-            info->mutable_instance_status()->set_status(grpc_instance_status_for(present_state));
-        }
+    bool snapshots_only = request->snapshots();
+    response.set_snapshots(snapshots_only);
 
-        auto vm_image = fetch_image_for(name, config->factory->fetch_type(), *config->vault);
-        auto original_release = vm_image.original_release;
+    auto populate_info = [&](VirtualMachine& vm, const std::shared_ptr<const Snapshot>& snapshot) {
+        auto* details = response.add_details();
+        if (snapshot)
+            populate_snapshot_info(vm, snapshot, details, have_mounts); // TODO@snapshots remove have_mounts
+        else
+            populate_instance_info(vm, details, request->no_runtime_information(), deleted, have_mounts);
+    };
 
-        if (!vm_image.id.empty() && original_release.empty())
+    auto process_snapshot_pick = [populate_info, snapshots_only](VirtualMachine& vm,
+                                                                 const SnapshotPick& snapshot_pick) {
+        for (const auto& snapshot_name : snapshot_pick.pick)
         {
-            try
-            {
-                auto vm_image_info = config->image_hosts.back()->info_for_full_hash(vm_image.id);
-                original_release = vm_image_info.release_title.toStdString();
-            }
-            catch (const std::exception& e)
-            {
-                mpl::log(mpl::Level::warning, category, fmt::format("Cannot fetch image information: {}", e.what()));
-            }
+            const auto snapshot = vm.get_snapshot(snapshot_name); // verify validity even if unused
+            if (!snapshot_pick.all_or_none || !snapshots_only)
+                populate_info(vm, snapshot);
         }
+    };
 
-        info->set_image_release(original_release);
-        info->set_id(vm_image.id);
-
-        auto vm_specs = vm_instance_specs[name];
-
-        auto mount_info = info->mutable_mount_info();
-
-        mount_info->set_longest_path_len(0);
+    auto fetch_detailed_report = [&](VirtualMachine& vm) {
```

Comment:
I prefer being explicit about what is being captured. I will say, if the list is very long, then using `&` is acceptable, but I think it's worthy of a comment for the code reader to know this is the case.

Reply:
OK, but this stems from the [`previous incarnation of Daemon::info`](https://github.com/canonical/multipass/blob/ef40bb593eb7db4e76969be303f54733598d2aa5/src/daemon/daemon.cpp#L1549). I will try to fix it too.

Reply:
Since this was already like this, then I don't think you need to spend time on fixing it then.